File: java/client/src/test/java/org/signal/libsignal/net/UnauthUsernamesServiceTest.kt
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.protocol.ServiceId.Aci
import org.signal.libsignal.protocol.util.Hex
import org.signal.libsignal.usernames.Username
import org.signal.libsignal.usernames.UsernameLinkInvalidEntropyDataLength
import org.signal.libsignal.usernames.UsernameLinkInvalidLinkData
import org.signal.libsignal.util.Base64
import java.nio.charset.StandardCharsets
import java.util.UUID
import kotlin.test.assertIs

class UnauthUsernamesServiceTest {
  companion object {
    val EXPECTED_USERNAME = "moxie.01"
    val ENCRYPTED_USERNAME =
      "kj5ah-VbEgjpfJsNt-Wto2H626DRmJSVpYPy0yPOXA8kiSFkBCD8ysFlJ-Z3MhiAnt_R3Nm7ZY0W5fiRDLVbhaE2z-KO2xdf5NcVbkewCzhvveecS3hHskDp1aSfbvwTZNNGPmAuKWvJ1MPdHzsF0w"
    val ENCRYPTED_USERNAME_ENTROPY =
      Hex.fromStringCondensedAssert(
        "4302c613c092a51c5394becffeb6f697300a605348e93f03c3db95e0b03d28f1",
      )
  }

  @Test
  fun testLookupUsernameHashSuccess() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val testHash = byteArrayOf(1, 2, 3, 4)
    val responseFuture = accountsService.lookUpUsernameHash(testHash)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    val expectedPath = "/v1/accounts/username_hash/" + Base64.encodeToStringUrl(testHash)
    assertEquals(expectedPath, request.pathAndQuery)

    // Send successful response with UUID
    val uuid = UUID.fromString("4FCFE887-A600-40CD-9AB7-FD2A695E9981")
    val jsonResponse =
      """
      {
        "uuid": "$uuid"
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<Aci?>>(result)
    assertEquals(Aci(uuid), successResult.result)
  }

  @Test
  fun testLookupUsernameHashNotFound() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val testHash = byteArrayOf(1, 2, 3, 4)
    val responseFuture = accountsService.lookUpUsernameHash(testHash)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    val expectedPath = "/v1/accounts/username_hash/" + Base64.encodeToStringUrl(testHash)
    assertEquals(expectedPath, request.pathAndQuery)

    // Send fake 404 response (user not found)
    fakeRemote.sendResponse(
      requestId,
      404,
      "Not Found",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<Aci?>>(result)
    assertEquals(null, successResult.result)
  }

  @Test
  fun testLookupUsernameHashRetryLater() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val testHash = byteArrayOf(1, 2, 3, 4)
    val responseFuture = accountsService.lookUpUsernameHash(testHash)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)

    // Send 429 response to trigger RetryLater
    fakeRemote.sendResponse(
      requestId,
      429,
      "Too Many Requests",
      arrayOf("retry-after: 120"),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val retryLater = assertIs<RequestResult.RetryableNetworkError>(result)
    assertEquals(120L, retryLater.retryAfter?.seconds)
  }

  @Test
  fun testLookupUsernameHashServerError() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val testHash = byteArrayOf(1, 2, 3, 4)
    val responseFuture = accountsService.lookUpUsernameHash(testHash)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)

    // Send server error response
    fakeRemote.sendResponse(
      requestId,
      500,
      "Internal Server Error",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val retryableNetworkError = assertIs<RequestResult.RetryableNetworkError>(result)
    assertIs<ServerSideErrorException>(retryableNetworkError.networkError)
  }

  @Test
  fun testLookupUsernameLinkSuccess() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val responseFuture = accountsService.lookUpUsernameLink(UUID(0, 0), ENCRYPTED_USERNAME_ENTROPY)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    assertEquals("/v1/accounts/username_link/00000000-0000-0000-0000-000000000000", request.pathAndQuery)

    // Send successful response
    val jsonResponse =
      """
      {
        "usernameLinkEncryptedValue": "$ENCRYPTED_USERNAME"
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<Username?>>(result)
    assertNotNull(successResult.result)
    assertEquals(EXPECTED_USERNAME, successResult.result!!.username)
  }

  @Test
  fun testLookupUsernameLinkNotFound() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val responseFuture = accountsService.lookUpUsernameLink(UUID(0, 0), ENCRYPTED_USERNAME_ENTROPY)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    assertEquals("/v1/accounts/username_link/00000000-0000-0000-0000-000000000000", request.pathAndQuery)

    // Send fake 404 response (link not found)
    fakeRemote.sendResponse(
      requestId,
      404,
      "Not Found",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<Username?>>(result)
    assertEquals(null, successResult.result)
  }

  @Test
  fun testLookupUsernameLinkGarbageCiphertext() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val responseFuture = accountsService.lookUpUsernameLink(UUID(0, 0), ENCRYPTED_USERNAME_ENTROPY)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    assertEquals("/v1/accounts/username_link/00000000-0000-0000-0000-000000000000", request.pathAndQuery)

    // Send successful response
    val jsonResponse =
      """
      {
        "usernameLinkEncryptedValue": "${ENCRYPTED_USERNAME}A"
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val failureResult = assertIs<RequestResult.NonSuccess<LookUpUsernameLinkFailure>>(result)
    assertIs<UsernameLinkInvalidLinkData>(failureResult.error)
  }

  @Test
  fun testLookupUsernameLinkServerError() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val responseFuture = accountsService.lookUpUsernameLink(UUID(0, 0), ENCRYPTED_USERNAME_ENTROPY)

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("GET", request.method)
    assertEquals("/v1/accounts/username_link/00000000-0000-0000-0000-000000000000", request.pathAndQuery)

    // Send server error response
    fakeRemote.sendResponse(
      requestId,
      500,
      "Internal Server Error",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val retryableNetworkError = assertIs<RequestResult.RetryableNetworkError>(result)
    assertIs<ServerSideErrorException>(retryableNetworkError.networkError)
  }

  @Test
  fun testLookupUsernameLinkBadEntropy() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val accountsService = UnauthUsernamesService(chat)
    val responseFuture =
      accountsService.lookUpUsernameLink(
        UUID(0, 0),
        ENCRYPTED_USERNAME_ENTROPY.copyOf(
          ENCRYPTED_USERNAME_ENTROPY.count() - 1,
        ),
      )

    val result = responseFuture.get()
    val failureResult = assertIs<RequestResult.NonSuccess<LookUpUsernameLinkFailure>>(result)
    assertIs<UsernameLinkInvalidEntropyDataLength>(failureResult.error)
  }
}

File: node/ts/test/chat/UnauthUsernamesServiceTest.ts
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { assert, config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';
import { Buffer } from 'node:buffer';
import * as uuid from 'uuid';

import * as Native from '../../Native.js';
import * as util from '../util.js';
import { TokioAsyncContext, UnauthUsernamesService } from '../../net.js';
import { connectUnauth } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';
import { Aci } from '../../Address.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('UnauthUsernamesService', () => {
  describe('lookUpUsernameHash', () => {
    it('can look up hashes', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const hash = Uint8Array.of(1, 2, 3, 4);
      const responseFuture = chat.lookUpUsernameHash({ hash });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        `/v1/accounts/username_hash/${Buffer.from(hash).toString('base64url')}`
      );

      const aci = '4fcfe887-a600-40cd-9ab7-fd2a695e9981';

      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            uuid: aci,
          })
        ),
      });

      const responseFromServer = await responseFuture;
      assert(responseFromServer !== null);
      assert(Aci.fromUuid(aci).isEqual(responseFromServer));
    });

    it('can look up unknown hashes', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const hash = Uint8Array.of(1, 2, 3, 4);
      const responseFuture = chat.lookUpUsernameHash({ hash });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        `/v1/accounts/username_hash/${Buffer.from(hash).toString('base64url')}`
      );

      fakeRemote.sendReplyTo(request, {
        status: 404,
        message: 'Not Found',
      });

      const responseFromServer = await responseFuture;
      assert.isNull(responseFromServer);
    });

    it('can handle challenge errors', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const hash = Uint8Array.of(1, 2, 3, 4);
      const responseFuture = chat.lookUpUsernameHash({ hash });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        `/v1/accounts/username_hash/${Buffer.from(hash).toString('base64url')}`
      );

      fakeRemote.sendReplyTo(request, {
        status: 428,
        message: 'Precondition Required',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            token: 'not-legal-tender',
            options: ['pushChallenge'],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.RateLimitChallengeError,
          token: 'not-legal-tender',
          options: new Set(['pushChallenge']),
        });
    });

    it('can handle server errors', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const hash = Uint8Array.of(1, 2, 3, 4);
      const responseFuture = chat.lookUpUsernameHash({ hash });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        `/v1/accounts/username_hash/${Buffer.from(hash).toString('base64url')}`
      );

      fakeRemote.sendReplyTo(request, {
        status: 500,
        message: 'Internal Server Error',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.IoError,
        });
    });
  });

  describe('lookUpUsernameLink', () => {
    const EXPECTED_USERNAME = 'moxie.01';
    const ENCRYPTED_USERNAME =
      'kj5ah-VbEgjpfJsNt-Wto2H626DRmJSVpYPy0yPOXA8kiSFkBCD8ysFlJ-Z3MhiAnt_R3Nm7ZY0W5fiRDLVbhaE2z-KO2xdf5NcVbkewCzhvveecS3hHskDp1aSfbvwTZNNGPmAuKWvJ1MPdHzsF0w';
    const ENCRYPTED_USERNAME_ENTROPY = Buffer.from(
      '4302c613c092a51c5394becffeb6f697300a605348e93f03c3db95e0b03d28f1',
      'hex'
    );

    it('can look up links', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: uuid.NIL,
        entropy: ENCRYPTED_USERNAME_ENTROPY,
      });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        '/v1/accounts/username_link/00000000-0000-0000-0000-000000000000'
      );

      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            usernameLinkEncryptedValue: ENCRYPTED_USERNAME,
          })
        ),
      });

      const responseFromServer = await responseFuture;
      assert.isNotNull(responseFromServer);
      assert.equal(responseFromServer.username, EXPECTED_USERNAME);
      assert.isNotEmpty(responseFromServer.hash);
    });

    it('can look up unknown links', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: uuid.NIL,
        entropy: ENCRYPTED_USERNAME_ENTROPY,
      });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        '/v1/accounts/username_link/00000000-0000-0000-0000-000000000000'
      );

      fakeRemote.sendReplyTo(request, {
        status: 404,
        message: 'Not Found',
      });

      const responseFromServer = await responseFuture;
      assert.isNull(responseFromServer);
    });

    it('can handle garbage ciphertexts', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: uuid.NIL,
        entropy: ENCRYPTED_USERNAME_ENTROPY,
      });

      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        '/v1/accounts/username_link/00000000-0000-0000-0000-000000000000'
      );

      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            usernameLinkEncryptedValue: `${ENCRYPTED_USERNAME}A`,
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.InvalidUsernameLinkEncryptedData,
        });
    });

    it('can handle server errors', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: uuid.NIL,
        entropy: ENCRYPTED_USERNAME_ENTROPY,
      });

      const request = await fakeRemote.assertReceiveIncomingRequest();
      assert(request !== null);

      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(
        '/v1/accounts/username_link/00000000-0000-0000-0000-000000000000'
      );

      fakeRemote.sendReplyTo(request, {
        status: 500,
        message: 'Internal Server Error',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.IoError,
        });
    });

    it('can handle bad UUIDs', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, _fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: 'not',
        entropy: ENCRYPTED_USERNAME_ENTROPY,
      });

      await expect(responseFuture).to.eventually.be.rejectedWith(TypeError);
    });

    it('can handle bad entropy', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, _fakeRemote] = connectUnauth<UnauthUsernamesService>(tokio);

      const responseFuture = chat.lookUpUsernameLink({
        uuid: uuid.NIL,
        entropy: ENCRYPTED_USERNAME_ENTROPY.subarray(1),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.InvalidEntropyDataLength,
        });
    });
  });
});
